base/services/w32time/CMakeLists.txt

```diff
@@ -0,0 +1,4 @@
+add_executable(W32Time W32Time.c ntpclient.c)
```

Comment:
Some tips for making this a dll.
The two main things you need is to change add_executable to add_library (with a SHARED parameter), and change set_module_type to use win32dll. You can see that in the other modules I mentioned, like shsvcs.

You then no longer need main/wmain/WinMain/wWinMain -- you'll need DllMain instead. Your DllMain doesn't need to do anything, so you can call DisableThreadLibraryCalls in DLL_PROCESS_ATTACH (lots of examples, just [grep for it](https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=DisableThreadLibraryCalls)) as an optimization.

Finally your dll needs to export functions (start with ServiceMain first, which you already have). To do that you add a .spec file listing all the functions you need to export and their parameters ([again, grep will have examples](https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=stdcall%20ServiceMain)). Then you add a spec2def line to CMakeLists.txt and add the generated .def file to the source files list (will have CMAKE_CURRENT_BINARY_DIR in the name, as seen in the example modules).

Then, to export W32TimeSyncNow, add a spec file entry for it and make sure your spec2def lines includes ADD_IMPORTLIB (this allows other modules to import it statically without using LoadLibrary/GetProcAddress). Then we can talk about the best way to call it from timedate.cpl.

Reply:
I reviewed your suggested services models and found that the shsvcs seemed to be the simplest out of the three based on named files, so I had decided to use it as a model to study. I have also researched the W32TimeSyncNow syntax (function definition) on-line and I think that I understand it. I will try and follow your tips above and see how far I can get. Thanks.

Reply:
If you want help, generally some devs are available on freenode in `#reactos-dev`.
Being the nature of irc, just ask your question (dont start like: can i ask a question?),
and someone should reply.
Not instant, because lots of people idle there.